### PR TITLE
Introduction of the Forward Conversion Tracker for ALICE 3

### PR DESCRIPTION
**Pull request by Cas1997**
Introduction of the Forward Conversion Tracker for ALICE 3.

Made by Cas1997

### CODE REVIEW

File: Common/SimConfig/src/SimConfig.cxx [hunk 1]
```diff
@@ -9,6 +9,8 @@
 // granted to it by virtue of its status as an Intergovernmental Organization
 // or submit itself to any jurisdiction.
 
+// Useless comment
```
Comment (mconcas) [inline Common/SimConfig/src/SimConfig.cxx]: remove

**Review by mconcas — DISMISSED** [review]
Hi @Cas1997, thanks for your PR.
For me there are no problems to include this in simulation for upgrades.
Despite activities for ALICE3 full simulation are now quite low because of the RUN3 focus (and the final LoI being submitted), I believe is valuable to keep the effort for your and all the possible future studies.

I am now pinging @sawenzel to ask what does he think about it.

File: Common/SimConfig/src/SimConfig.cxx (hunk 1 above)
Comment (Cas1997) [reply Common/SimConfig/src/SimConfig.cxx]: Oh, that must have slipped my check

File: Detectors/Base/src/Aligner.cxx [hunk 2]
```diff
@@ -24,7 +24,8 @@ using DetID = o2::detectors::DetID;
 
 DetID::mask_t Aligner::getDetectorsMask() const
 {
-  return DetID::getMask(mDetectors) & (~DetID::getMask(DetID::CTP));
+  return DetID::getMask(mDetectors) & (~DetID::getMask(DetID::CTP)) & (~DetID::getMask(DetID::FCT));
```
Comment (benedikt-voelkel) [inline Detectors/Base/src/Aligner.cxx]: The FCT needs to go here and it needs to be made sure that we get the alignment properly from the CCDB for FCT as well. (as you might remember @Cas1997 I only suggested that as a quick workaround ;) )

How would we approach that @shahor02 ?

File: macro/duplicateHits.C [hunk 3]
```diff
@@ -23,6 +23,7 @@
 
 #ifdef ENABLE_UPGRADES
 #include "EndCapsSimulation/Hit.h"
+#include "ForwardTrackerSimulation/Hit.h"
```
Comment (benedikt-voelkel) [inline macro/duplicateHits.C]: why would that be needed?

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+ClassImp(FCTLayer);
+
+FCTLayer::~FCTLayer() = default;
+
+FCTLayer::FCTLayer(Int_t layerDirection, Int_t layerNumber, std::string layerName, Float_t z, Float_t rIn, Float_t rOut_SideL, Float_t Layerx2X0, Int_t type)
+{
+  // Creates a simple parametrized EndCap layer covering the given
+  // pseudorapidity range at the z layer position
+  mDirection = layerDirection;
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: you could do that like
```c++
FCTLayer::FCTLayer(...) : mDirection{layerDirection}, mLayerNumber{layerNumber}, ...
```

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+class TGeoMedium;
+
+using namespace TMath;
+using namespace o2::fct;
+using namespace o2::itsmft;
+
+ClassImp(FCTLayer);
+
+FCTLayer::~FCTLayer() = default;
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx:40]: I reckon that should go to the header file

File: macro/duplicateHits.C (hunk 3 above)
Comment (Cas1997) [reply macro/duplicateHits.C]: I have no clue if I am being honest. I copied the O2 around two months ago and then this might have been there and a later pull request might have removed it (and so this older version is trying to place it back). I will look for the pull request which removed it.

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    TGeoVolume* layerVol = new TGeoVolume(mLayerName.c_str(), LayerComp, medAir);
+
+    LOG(info) << "Inserting " << sensVol->GetName() << " inside " << chipVol->GetName();
+    chipVol->AddNode(sensVol, 1, nullptr);
+
+    LOG(info) << "Inserting " << chipVol->GetName() << " inside " << layerVol->GetName();
+    layerVol->AddNode(chipVol, 1, nullptr);
+
+    // Finally put everything in the mother volume
+    auto* FwdLayerRotation = new TGeoRotation("FwdDiskRotation", 0, 0, 180);
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: you can omit the `*`

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    TGeoCompositeShape* LayerComp = new TGeoCompositeShape("SensorComp", "LayerBox - LayerTube");
+
+    TGeoMedium* medSi = gGeoManager->GetMedium("FCT_SI$");
+    TGeoMedium* medAir = gGeoManager->GetMedium("FCT_AIR$");
+
+    TGeoVolume* sensVol = new TGeoVolume(sensName.c_str(), SensorComp, medSi);
+    TGeoVolume* chipVol = new TGeoVolume(chipName.c_str(), ChipComp, medSi);
+    TGeoVolume* layerVol = new TGeoVolume(mLayerName.c_str(), LayerComp, medAir);
+
+    LOG(info) << "Inserting " << sensVol->GetName() << " inside " << chipVol->GetName();
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: actually there are a lot of `LOG(info)` in this file and others. Could some of them be replaced by a `LOG(debug)`?

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+#include <TGeoMatrix.h>         // for TGeoCombiTrans, TGeoRotation, etc
+#include <TGeoTube.h>           // for TGeoTube, TGeoTubeSeg
+#include <TGeoVolume.h>         // for TGeoVolume, TGeoVolumeAssembly
+#include <TGeoCompositeShape.h> // for TGeoCompositeShape
+#include "TMathBase.h"          // for Abs
+#include <TMath.h>              // for Sin, RadToDeg, DegToRad, Cos, Tan, etc
+
+#include <cstdio> // for snprintf
+
+class TGeoMedium;
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: why is that here?
Either you pull in the header or you don't I would say. I think right now you are "lucky" that `TGeoMedium` is actually included by something else

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+#include "TCanvas.h"
+#include "TDatime.h" // for GetDate, GetTime
+
+#include <cstdio> // for NULL, snprintf
+
+class FairModule;
+
+class TGeoMedium;
+
+class TParticle;
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: same comment as for `FCTLayer`

File: Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/FCTLayer.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file FCTLayer.h
+/// \brief Definition of the FCTLayer class
+
+#ifndef ALICEO2_FCT_UPGRADEV3LAYER_H_
+#define ALICEO2_FCT_UPGRADEV3LAYER_H_
+
+#include <TGeoManager.h>            // for gGeoManager
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/FCTLayer.h]: that include shouldn't be needed here
do it in the source file if needed there

File: Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+  void BeginPrimary() override { ; }
+  void PostTrack() override { ; }
+  void PreTrack() override { ; }
+
+  /// Returns the number of layers
+  Int_t getNumberOfLayers() const { return mNumberOfLayers; }
+
+  void buildBasicFCT(const FCTBaseParam& param);
+  void buildFCTV1();
+  void buildFCTFromFile(std::string);
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h:116]: that could be `const&` as far as I can see from the implementation of this method

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx [hunk 12]
```diff
@@ -0,0 +1,592 @@
...
+
+  for (Int_t layerNumber = 0; layerNumber < mNumberOfLayers; layerNumber++) {
+    std::string layerName = GeometryTGeo::getFCTLayerPattern() + std::to_string(layerNumber); // + mNumberOfLayers * direction);
+    mLayerName.push_back(layerName);
+
+    // Adds evenly spaced layers
+    Float_t layerZ = z_first + (layerNumber * z_length / (mNumberOfLayers - 1)) * std::copysign(1, z_first);
+    Float_t rIn = std::abs(layerZ * std::tan(2.f * std::atan(std::exp(-etaIn))));
+    Float_t rOut = std::abs(layerZ * std::tan(2.f * std::atan(std::exp(-etaOut))));
+    auto& thisLayer = mBackwardLayers.emplace_back(0, layerNumber, layerName, layerZ, rIn, rOut, Layerx2X0, type);
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: you are not using `thisLayer`.
Only `mBackwardLayers.emplace_back(0, layerNumber, layerName, layerZ, rIn, rOut, Layerx2X0, type);` will do

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx [hunk 13]
```diff
@@ -0,0 +1,592 @@
...
+    mLayerName.push_back(layerName);
+    Float_t z = layersConfig[layerNumber][0];
+
+    Float_t rIn = layersConfig[layerNumber][1];
+    Float_t rOut = layersConfig[layerNumber][2];
+    Float_t x0 = layersConfig[layerNumber][3];
+
+    LOG(info) << "Adding Layer " << layerName << " at z = " << z;
+    // Add layers
+    auto& thisLayer = mBackwardLayers.emplace_back(0, layerNumber, layerName, z, rIn, rOut, x0, type);
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: you are not using `thisLayer`.
Only `mBackwardLayers.emplace_back(0, layerNumber, layerName, layerZ, rIn, rOut, Layerx2X0, type);` will do

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx [hunk 14]
```diff
@@ -0,0 +1,592 @@
...
+    {-490.0, 6.6, 48.9, layersx2X0}};
+
+  mLayerID.clear();
+  mLayerName.clear();
+  mBackwardLayers.clear();
+
+  Int_t type = 0; // Disk
+
+  for (int layerNumber = 0; layerNumber < mNumberOfLayers; layerNumber++) {
+    std::string directionName = std::to_string(0);
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: why are you defining this variable? Is it supposed to be actually used? Right now it looks like you could just add a "0" at the next line where you use it

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+    LOG(info) << "Inserting " << layerVol->GetName() << " inside " << motherVolume->GetName();
+    motherVolume->AddNode(layerVol, 1, FwdLayerCombiTrans);
+
+    return;
+  }
+}
+
+void FCTLayer::createConverterLayer(TGeoVolume* motherVolume)
+{
+  if (mLayerNumber >= 0) {
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: just an alternative proposal which sometimes can make the code a bit cleaner: return as early as possible
```c++
if (mLayerNumber < 0) {
  return;
}
// rest of the code

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+    layerVol->AddNode(chipVol, 1, nullptr);
+
+    // Finally put everything in the mother volume
+    auto* FwdDiskRotation = new TGeoRotation("FwdDiskRotation", 0, 0, 180);
+    auto* FwdDiskCombiTrans = new TGeoCombiTrans(0, 0, mZ, FwdDiskRotation);
+
+    LOG(info) << "Inserting " << layerVol->GetName() << " inside " << motherVolume->GetName();
+    motherVolume->AddNode(layerVol, 1, FwdDiskCombiTrans);
+  }
+  return;
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: not needed

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx (hunk 14 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: The FT3, the code the FCT was based on, has backward layers, "0", and forward layers, "1". When I wanted to get rid of this a conflict occurred in finding the layers when they were later called by name. This is hopefully resolved as of now since I moved the initialization statements in the Detector::Detector class to Detector::ConstructGeometry. As I learned, a dummy detector object is created when the detector is called. Thus I moved this to the ConstructGeometry class. This is not done yet in the FT3 and a pull request should be created to fix that there too.

In short: For the naming of the layers, yes, a "0" does get added, but as far as I know it has no function right now. This is a remnant from the code for the FT3

File: Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+
+  void exportLayout();
+
+ protected:
+  std::vector<Int_t> mLayerID;
+  std::vector<Int_t> mConverterLayerId;
+  std::vector<TString> mLayerName;
+  std::vector<TString> mConverterLayerName;
+  Int_t mNumberOfLayers;
+  Int_t mNumberOfConverterLayers;
```
Comment (benedikt-voelkel) [inline Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h]: is it your expectation that somebody derives from this class or why are these members protected and not private?

File: Detectors/Base/src/Aligner.cxx (hunk 2 above)
Comment (shahor02) [reply Detectors/Base/src/Aligner.cxx]: Since you don't have alignable volumes at the moment, just a default (empty) placeholder object is enough, I've just uploaded it to http://alice-ccdb.cern.ch/browse/FCT/Calib/Align.
The defaults object is generated like in https://github.com/AliceO2Group/AliceO2/blob/dev/macro/UploadDummyAlignment.C (which would cover also FCT once this PR is merged, I've uploaded the object manually yo avoid re-generation of defaults for all detectors).
So, just remove this hack.

Comment (benedikt-voelkel) [reply Detectors/Base/src/Aligner.cxx]: Thanks a lot!

@Cas1997 I think we can then move ahead. Once you proposed all changes we can re-iterate your PR :)

Comment (Cas1997) [reply Detectors/Base/src/Aligner.cxx]: Done

File: macro/duplicateHits.C (hunk 3 above)
Comment (Cas1997) [reply macro/duplicateHits.C]: Removed

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx (hunk 4 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: Done

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx (hunk 5 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx:40]: I tried, but the building failed afterwards.

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx (hunk 6 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: Done

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx (hunk 8 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: Resolved

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx (hunk 9 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: Resolved

File: Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/FCTLayer.h (hunk 10 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/FCTLayer.h]: Resolved

File: Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h (hunk 11 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h:116]: I tried, but that returned an error so I left it as it is

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx (hunk 12 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: Done

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx (hunk 13 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: Done

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx (hunk 14 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: Resolved. This has its own commit: https://github.com/AliceO2Group/AliceO2/pull/8679/commits/ccdaf8c36ad4cc11041ce94bd33727d616766e0f

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx (hunk 15 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: Done. Thanks for the tip. It does indeed make it more readable

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx (hunk 16 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/FCTLayer.cxx]: Removed

File: Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h (hunk 17 above)
Comment (Cas1997) [reply Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h]: Made private. No derivation is expected

File: Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h (hunk 11 above)
Comment (benedikt-voelkel) [reply Detectors/Upgrades/ALICE3/FCT/simulation/include/FCTSimulation/Detector.h:116]: well, you need to change that here and in the source file of course. from the source file (where the actual implementation of the function is) I don't see why this cannot be done.
Note: The function signatures have to match, so you cannot have `std::string` in one place and `std::string const&` in the other

File: Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx (hunk 14 above)
Comment (benedikt-voelkel) [reply Detectors/Upgrades/ALICE3/FCT/simulation/src/Detector.cxx]: I see, thanks!
Just as an advice maybe: A line like this - regardless of whether this code was adapted from other code - looks weird if there is nothing else done other than just making a string from a number

**Comment by benedikt-voelkel** [comment]
Thanks for addressing all the comments @Cas1997 
I think we can move further towards merging. @sawenzel might want to have another look as well

**Comment by benedikt-voelkel** [comment]
Hej @TimoWilken 
The ALICE CIs seem not to have started, however the last commit was added to the source branch more than 5 days ago. Is there any reason or are they just still queuing?

**Comment by TimoWilken** [comment]
Hey @benedikt-voelkel, the CI doesn't run by default on a user's first pull request to O2. This PR needs an approving review for the CI to start.

**Comment by benedikt-voelkel** [comment]
ok, I see, thanks for the explanation!

**Review by sawenzel — DISMISSED** [review]

**Comment by sawenzel** [comment]
Approved to trigger CI

**Comment by sawenzel** [comment]
@Cas1997 : I think the code looks fine to me but I have a comment on the meta-content of the git commits.
It would be nice and appropriate to have real names and institutional information in the git commits. For the moment, I see for instance:
```
commit 5af539e2e3a03aeb324e6218d47718b0738dff50
Author: Cas1997 <casvv4@gmail.com>
```
which is rather cryptic. Please use full names here, which makes identifying the author of a piece of code much easier.
In principle this is to be set in a global .gitconfig file.

Would you be able to do this (and possible squash all commits together in a single well-contained one)?

**Comment by Cas1997** [comment]
@sawenzel I understand your comment. I will change my name and squash the commits.

**Comment by sawenzel** [comment]
@Cas1997 : Thanks for the squash... but the name and email information seems still be the same as before.

**Review by sawenzel — DISMISSED** [review]

**Comment by Cas1997** [comment]
@sawenzel I must admit, that is a bit awkward. I thought I had changed it, but it appears not to be the case

**Comment by sawenzel** [comment]
You might have to add some special treatment to `git commit --amend` to overwrite the initial author.

**Comment by sawenzel** [comment]
Something like: `git commit --amend --author="John Doe <john@doe.org>"` ... but for the future commits you should be safe with a .gitconfig.

**Comment by Cas1997** [comment]
@sawenzel How about now? I think I got it right this time

**Review by sawenzel — APPROVED** [review]

**Comment by benedikt-voelkel** [comment]
Dear @Cas1997 
The workflow for the formatting CI changed only 4 days ago. This is potentially related to the fact that now the CI is failing for `run/SimExamples/SimAsService_biasing1/duplicateHits.macro`.
There should be trailing whitespaces. Could you remove them and push again?

Concerning the failing CI `build/O2/o2-dataflow-cs8` it seems to be unrelated to your PR I would say

```bash
## sw/BUILD/QualityControl-latest/log
--
2022-05-23 06:58:26.639729     task : qc/TST/MO/functional_test/object1
2022-05-23 06:58:26.639760     task : qc/TST/MO/functional_test/object2
2022-05-23 06:58:26.639768     task : qc/TST/MO/functional_test/path
/sw/SOURCES/QualityControl/v1.59.1/v1.59.1/Framework/test/testDbFactory.cxx(102): [1;31;49merror: in "db_ccdb_listing": check std::find(objectNames.begin(), objectNames.end(), "/object1") != objectNames.end() has failed[0;39;49m

[1;31;49m*** 1 failure is detected in the test module "DbFactory test"
```
